crates/catalog/src/mutator.rs

```diff
@@ -30,12 +30,47 @@ impl CatalogMutator {
         self.client.as_ref()
     }
 
+    /// Commit the catalog state.
+    /// This persists the state to the metastore.
+    /// The `current_catalog_version` is the version of the catalog prior to the state being committed.
+    /// This is usually `state.version - 1`, but not guaranteed.
+    /// This will retry commits if we were working with an out of date catalog.
+    pub async fn commit_state(
+        &self,
+        current_catalog_version: u64,
+        state: CatalogState,
+    ) -> Result<Arc<CatalogState>> {
+        let client = match &self.client {
+            Some(client) => client,
+            None => return Err(CatalogError::new("metastore client not configured")),
+        };
+
+        match client
+            .commit_state(current_catalog_version, state.clone())
+            .await
+        {
+            Err(CatalogError {
+                msg,
+                strategy: Some(ResolveErrorStrategy::FetchCatalogAndRetry),
+            }) => {
+                debug!(error_message = msg, "retrying commit");
+                // refresh the cached state and retry the commit
+                client.refresh_cached_state().await?;
+
+                client.commit_state(state.version, state).await
```

Comment:
I think I don't understand how this retry helps.

If we get here it's probably because the _service_ failed to commit, because the state The version in `state` hasn't changed since the last attempt, the version in `state`is likely to be _more_ wrong now. 

I think the desired flow for this is: 

- hand the client the state that you want to commit and the version of the catalog that you modified to get the new state. 
- if we get an error that should be retried, discard the state, reapply the mutation to the _new_ state of the catalog fetched after the last failure, and attempt to commit this state. 

Right now, based on my reading of the code, I think: 
- we get a failure from the service, presumably because the new state of the catalog is higher than the state that we're trying to apply
- we apply the same state, and the client rejects before sending it to the service, it because it's stale (this is correct, but it means that the error will either:
  - always bubble up to the user.
  - recursively retry forever.

Reply:
no recursion takes place here. It only tries to recommit it once. We can pretty easily remove this, it was added as the `mutation` path had this same retry logic and didn't want to deviate too much from the original code.